akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -303,6 +224,52 @@ private[http] object StreamUtils {
 
     def runInContext(block: ⇒ Unit): Unit = getAsyncCallback[AnyRef](_ ⇒ block).invoke(null)
   }
+
+  private val EmptySource = Source.empty
+
+  /** Dummy name to signify that the caller asserts that cancelSource is only run from within a GraphInterpreter context */
+  val OnlyRunInGraphInterpreterContext: Materializer = null
+  /**
+   * Tries to guess whether a source needs to cancelled and how. In the best case no materialization should be needed.
+   */
+  def cancelSource(source: Source[_, _])(implicit materializer: Materializer): Unit = source match {
+    case EmptySource ⇒ // nothing to do with empty source
+    case x ⇒
+      val mat =
+        GraphInterpreter.currentInterpreterOrNull match {
+          case null if materializer ne null ⇒ materializer
+          case null                         ⇒ throw new IllegalStateException("Need to pass materializer to cancelSource if not run from GraphInterpreter context.")
+          case x                            ⇒ x.subFusingMaterializer // try to use fuse if already running in interpreter context
+        }
+      x.runWith(Sink.ignore)(mat)
+  }
```

Comment:
I wonder if my weird materializer could be used here actually once it's out there...